Modules/_datetimemodule.c

```diff
@@ -2633,6 +2825,46 @@ date_fromordinal(PyObject *cls, PyObject *args)
     return result;
 }
 
+/* Return the new date from a string as generated by date.isoformat() */
+static PyObject *
+date_fromisoformat(PyObject *cls, PyObject *dtstr) {
+    assert(dtstr != NULL);
+
+    if (!PyUnicode_Check(dtstr)) {
+        PyErr_SetString(PyExc_TypeError, "fromisoformat: argument must be str");
+        return NULL;
+    }
+
+    Py_ssize_t len;
+
+    const char * dt_ptr = PyUnicode_AsUTF8AndSize(dtstr, &len);
+
+    int year = 0, month = 0, day = 0;
+
+    int rv;
+    if (len == 10) {
+        rv = parse_isoformat_date(dt_ptr, &year, &month, &day);
+    } else {
+        rv = -1;
+    }
+
+    if (rv < 0) {
+        PyErr_Format(PyExc_ValueError, "Invalid isoformat string: %s",
+                     dt_ptr);
+        return NULL;
+    }
+
+    PyObject *result;
+    if (PyDate_CheckExact(cls)) {
```

Comment:
This check is actually wrong, so this branch is never reached. The correct code is:

```python
if ( (PyTypeObject*)cls == &PyDateTime_DateType ) {
```